manimlib/utils/tex.py

```diff
@@ -0,0 +1,258 @@
+            tpl += self.ctex_preamble_text
+        else:
+            tpl += self.tex_preamble_text
+
+        tpl += "\n" r"\begin{document}" "\n"
+        tpl += f"\n{self.text_to_replace}\n"
+        tpl += "\n" r"\end{document}" 
+
+        self.template_text=tpl
+        
+    def prepend_package(self, pkg):
+        """Adds a new package (or several new packages) 
+        before all other packages. Sometimes, the order of 
+        the `\\usepackage` directives is relevant.
+        
+        Parameters
+        ----------
+        pkg : :class:`str`
+            The package name, e.g. "siunitx"
+        """
+        self.common_packages.insert(0, pkg)
+        self.rebuild_cache()
+
+    def append_package(self, pkg):
+        """Adds a new package (or several new packages) 
+        after all other packages. Sometimes, the order of 
+        the `\\usepackage` directives is relevant.
+        
+        Parameters
+        ----------
+        pkg : :class:`str`
+            The package name, e.g. "siunitx"
+        """
+        self.common_packages.append(pkg)
+        self.rebuild_cache()
+
+    def append_to_preamble(self,text):
+        """Adds commands (e.g. macro definitions) at the end of the preamble.
+        
+        Parameters
+        ----------
+        text : :class:`str`
+            The text to be included, e.g. "\\newcommand{\\R}{\\mathbb{Q}}"
+        """
+        if self.use_ctex:
+            self.ctex_preamble_text += text
+        else:
+            self.tex_preamble_text += text
+        self.rebuild_cache()
+        pass
+
+    def clear_preamble(self):
+        """Removes custom definitions from the LaTeX preamble.
+        This does not affect the imported packages or documentclass."""
+        self.common_preamble_text = ""
+        self.ctex_preamble_text = ""
+        self.tex_preamble_text = ""
+        self.rebuild_cache()
+        pass
+
+    def generate_simple_tex_command_with_options(self,command,params):
+        """
+        Function for creating LaTeX command strings with or without options.
+        Internally used to generate `\\usepackage{...}`
+        
+        Parameters
+        ----------
+        command : :class:`str`
+            The command, e.g. `"usepackage"`
+        params : Iterable[Iterable[Union[:class:`str`, Iterable[:class:`str`]]]]
+            The options, e.g. `["packagename",["option1","option2"]]`
+        
+        Returns
+        -------
+        :class:`str`
+            The generated command.
+        """
+        if isinstance(params,list):
+            line = "\\" \
+                   + command \
+                   + r"[" \
+                   + ",".join(params[1]) \
+                   + r"]{" \
+                   + params[0] \
+                   + r"}" + "\n"
+        else:
+            line = "\\" \
+                   + command \
+                   + r"{" \
+                   + params \
+                   + r"}" + "\n"
+        return line
```

Comment:
Smaller name, also should allow any number of params.

Note here that `required_params` would represent one `{}` each, while each in `optional_params` is separated by a comma in a single `[]`.

```suggestion
    import collections  # can (and probably should) move this to the beginning of the file as well
    def generate_tex_command(self,command, *, required_params, optional_params):
        """
        Function for creating LaTeX command strings with or without options.
        Internally used to generate `\\usepackage{...}`
        
        Parameters
        ----------
        command : :class:`str`
            The command, e.g. `"usepackage"`
        required_params : Iterable[:class:`str`]
            The required parameters of this command, each wrapped in `{}`s.
        optional_params : Iterable[:class:`str`]
             The optional parameters of this command, each separated by a comma inside one `[]`.
        
        Examples
        --------
        ::
            generate_tex_command("usepackage", required_params=["packagename"], optional_params=["option1", "option2"])
        
        Returns
        -------
        :class:`str`
            The generated command.
        """
        optional_params = list(optional_params)  # so we can measure its length
        return r"\{0}{1}{2}".format(
            command,
            f"[{','.join(optional_params)}]" if optional_params else "",
            "".join("{" + param + "}" for param in required_params)
        )
```

Reply:
This is, of course, a nice improvement. When I wrote it, the function was just meant to be for internal use, so I did not see a need for such a versatile solution, hence the word _simple_ in the name.